consensus/polybft/validators_snapshot.go

```diff
@@ -36,127 +54,106 @@ func (v *validatorsSnapshotCache) GetSnapshot(blockNumber uint64, parents []*typ
-		initialSnapshot, err := v.computeSnapshot(nil, 0, parents)
+		// log.Trace("Building validators snapshot from scratch...")
+		genesisBlockSnapshot, err := v.computeSnapshot(nil, 0, parents)
 		if err != nil {
 			return nil, fmt.Errorf("failed to compute snapshot for epoch 0: %w", err)
 		}
 
-		snapshot = initialSnapshot
-		// store same snapshot for epochs 0 and 1, since it is shared between those two epochs
-		// (and there might be use case to require snapshot for the genesis block, i.e. epoch 0)
-		err = v.storeSnapshot(0, initialSnapshot)
+		err = v.storeSnapshot(genesisBlockSnapshot)
 		if err != nil {
 			return nil, fmt.Errorf("failed to store validators snapshot for epoch 0: %w", err)
 		}
 
-		v.logger.Trace("Built validators snapshot")
+		latestValidatorSnapshot = genesisBlockSnapshot
+
+		v.logger.Trace("Built validators snapshot for genesis block")
 	}
 
-	// Create most recent snapshot by incrementally applying validator set deltas from the resolved block
-	// which have snapshot associated to the last block which is less than the one snapshot is requested for.
 	deltasCount := 0
 
-	v.logger.Trace("Applying deltas started...",
-		"lastBlockWithSnapshot", lastBlockWithSnapshot,
-		"blockNumber", blockNumber)
-
-	for lastBlockWithSnapshot < blockNumber {
-		lastBlockWithSnapshot++
-
-		v.logger.Trace("Applying delta", "blockWithSnapshot", lastBlockWithSnapshot)
+	v.logger.Trace("Applying deltas started...", "LatestSnapshotEpoch", latestValidatorSnapshot.Epoch)
 
-		intermediateSnapshot, err := v.computeSnapshot(snapshot, lastBlockWithSnapshot, parents)
+	// Create the snapshot for the desired block (epoch) by incrementally applying deltas to the latest stored snapshot
+	for latestValidatorSnapshot.Epoch < epochToGetSnapshot {
+		nextEpochEndBlockNumber, err := v.getNextEpochEndingBlock(latestValidatorSnapshot.EpochEndingBlock)
 		if err != nil {
-			return nil, fmt.Errorf("failed to compute snapshot for block %d: %w", lastBlockWithSnapshot, err)
+			return nil, fmt.Errorf("failed to get the epoch ending block for epoch: %d. Error: %w",
+				latestValidatorSnapshot.Epoch+1, err)
 		}
 
-		snapshot = intermediateSnapshot
-		deltasCount++
-	}
+		v.logger.Trace("Applying delta", "epochEndBlock", nextEpochEndBlockNumber)
 
-	if deltasCount > 0 {
-		err := v.storeSnapshot(lastBlockWithSnapshot, snapshot)
+		intermediateSnapshot, err := v.computeSnapshot(latestValidatorSnapshot, nextEpochEndBlockNumber, parents)
 		if err != nil {
-			return nil, fmt.Errorf("failed to store validators snapshot for block %d: %w", lastBlockWithSnapshot, err)
+			return nil, fmt.Errorf("failed to compute snapshot for epoch %d: %w", latestValidatorSnapshot.Epoch+1, err)
 		}
+
+		latestValidatorSnapshot = intermediateSnapshot
+		if err = v.storeSnapshot(latestValidatorSnapshot); err != nil {
+			return nil, fmt.Errorf("failed to store validators snapshot for epoch %d: %w", latestValidatorSnapshot.Epoch, err)
+		}
```

Comment:
Couldn't you done this outside of the for loop? Not sure why you need to store all those intermediary snapshots persisted?

Reply:
I'd say that the odds are pretty low that we are going to need some of the past snapshots (most of the time we need the latest one or the one before it). Also when saving the most recent one only, it is good enough, because it would represent a starting point if there should be built some more recent snapshot.

I agree that persistence of snapshot is "cheaper" than calculation of snapshots, but to me it is really a question whether we are ever need those intermediary snapshots.

Reply:
I agree. That is something we can discuss further. To me it's better safe then sorry. Save it, let it stay in cache and db, it's not a huge amount of data we are saving. Plus, we have that cleaning mechanism, so it won't stay in cache and db for too long.